File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(head == null || head.next == null || left == right){
            return head;
        }
        
        ListNode sentinal = new ListNode(0, head);
        
        ListNode tail = sentinal;
        for(int i = 1; i < left; i++){
            tail = tail.next;
        }
        
        ListNode prev = tail.next, curr = tail.next.next;
        for(int i = 0; i < right - left; i++){
            ListNode ncurr = curr.next;
            curr.next = prev;
            prev = curr;
            curr = ncurr;
        }
        tail.next.next = curr;
        tail.next = prev;
        
        return sentinal.next;
    }
}

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right or not head or not head.next:
            return head
        
        sentinal = ListNode(0, head)
        
        ltail = sentinal
        for i in range(left - 1):
            ltail = ltail.next
        
        prev, curr = ltail.next, ltail.next.next
        for _ in range(right - left):
            ncurr, curr.next = curr.next, prev
            prev, curr = curr, ncurr
        ltail.next.next = curr
        ltail.next = prev
        
        return sentinal.next